src/vm/types.rs

```diff
@@ -425,8 +422,18 @@ impl AtomTypeIdentifier {
 
     fn admits(&self, other: &AtomTypeIdentifier) -> bool {
         match self {
-            AtomTypeIdentifier::AnyType => {
-                true
+            AtomTypeIdentifier::OptionalType(ref my_inner_type) => {
+                if let AtomTypeIdentifier::OptionalType(other_inner_type) = other {
+                    // Option types will always admit a "NoType" OptionalType -- which
+                    //   can only be a None
+                    if other_inner_type.is_no_type() {
```

Comment:
Should we also have this behavior when `my_inner_type.is_no_type()  == true`?

Reply:
No, I don't think so -- we wouldn't want something that's typed as `(optional notype)` to admit anything other than _maybe_ `(optional notype)`, since you wouldn't be able to tell via static analysis what the eventual type would be.